src/nanoemoji/write_glyphgraph.py

```diff
@@ -0,0 +1,189 @@
+        new_edge = (src, dest) not in self.edges
+        if src is not None and new_edge:
+            self.graph.edge(src, dest)
+        self.edges.add((src, dest))
+        return new_edge
+
+
+def _base_glyphs(font, filter_fn):
+    for base_glyph in font["COLR"].table.BaseGlyphV1List.BaseGlyphV1Record:
+        if filter_fn(base_glyph):
+            yield base_glyph
+
+
+def _only(seq):
+    seq = tuple(seq)
+    if len(seq) != 1:
+        raise ValueError("Need 1 entry, got " + len(seq))
+    return seq[0]
+
+
+def _indent(depth):
+    return depth * "  "
+
+
+def _color_line_node_id(palette, color_line):
+    id_parts = ["ColorLine", color_line.Extend.name]
+    for stop in color_line.ColorStop:
+        cpal_color = palette[stop.Color.PaletteIndex]
+        cpal_color = cpal_color._replace(alpha=int(cpal_color.alpha * stop.Color.Alpha.value))
+        id_parts.append(f"{cpal_color.hex()}@{stop.StopOffset.value:.1f}")
+    return "_".join(id_parts)
+
+def _paint_node_id(palette, paint):
+    if paint.Format == 1:
+        cpal_color = palette[paint.Color.PaletteIndex]
+        colr_alpha = paint.Color.Alpha.value
+        return f"Solid_{cpal_color}_{colr_alpha:.4f}"
+    if paint.Format == 2:
+        id_parts = (
+            "Linear",
+            f"p0({paint.x0.value}, {paint.y0.value})",
+            f"p1({paint.x1.value}, {paint.y1.value})",
+            f"p2({paint.x2.value}, {paint.y2.value})",
+            _color_line_node_id(palette, paint.ColorLine),
+            )
+        return "_".join(id_parts)
+    if paint.Format == 3:
+        id_parts = (
+            "Radial",
+            f"c0({paint.x0.value}, {paint.y0.value})",
+            f"r0 {paint.r0.value}",
+            f"c1({paint.x1.value}, {paint.y1.value})",
+            f"r1 {paint.r1.value}",
+            _color_line_node_id(palette, paint.ColorLine),
+            )
+        return "_".join(id_parts)
+    if paint.Format == 4:
+        return "Glyph_" + paint.Glyph
+    if paint.Format == 5:
+        id_parts = (
+            "Slice",
+            "Base",
+            paint.Glyph,
+            "%d..%d" % (paint.FirstLayerIndex, paint.LastLayerIndex),
+            )
+        return "_".join(id_parts)
+    if paint.Format == 6:
+        id_parts = (
+            "Transform",
+            f"î {paint.Transform.xx.value},{paint.Transform.xy.value}",
+            f"ĵ {paint.Transform.yx.value},{paint.Transform.yy.value}",
+            f"dx {paint.Transform.dx.value}",
+            f"dy {paint.Transform.dy.value}",
+            _paint_node_id(palette, paint.Paint),
+            )
+        return "_".join(id_parts)
+    if paint.Format == 7:
+        id_parts = (
+            "Composite",
+            _paint_node_id(palette, paint.SourcePaint),
+            paint.CompositeMode.name,
+            _paint_node_id(palette, paint.BackdropPaint),
+            )
+        return "_".join(id_parts)
+    raise NotImplementedError(f"id for format {paint.Format} ({dir(paint)})")
```

Comment:
we need an elif paint.Format == 8 for making a digraph with PaintColrLayers

Reply:
Oh I have that; must just not have uploaded it